### ISSUE TITLE
Set Window title and Tab title separately in iTerm2, add bell char

### ISSUE THREAD

**Issue by eugenesvk** [fish-shell/fish-shell#2692]
In `zsh` I am able to set tab and window titles separately in iTerm2, e.g. I have
1) __Tab__ title: current folder only to easily differentiate tabs by which folder they're open at. In path  `~` replaces `username` (`echo -ne "\033]1;{$PWD1/#username/~}\007"`, where `PWD1={$PWD##*/}`
2) __Win__ title: full path with `~` instead of `/users/username` (`echo -ne "\033]2;{$PWD/#"$HOME"/~}\007"`)

As I understand from https://github.com/fish-shell/fish-shell/issues/47, `fish` just sets both windows and tab titles with once `fish_title` function.
Is it possible to separate those two, maybe have a separate `tab_title` function or something so that users who are not using it continue to get both windows and tab title within the old function?
Also, in `zsh` I am able to add a bell char (`\007`) to the titles. How do I do that in `fish`?
Thanks!

P.S.
My system is `OS X 10.11.2`, `iTerm2 2.9.20160113` and `fish 2.2.0-542-g4b2495f`

**Comment by floam** [comment]
For me in both iTerm2 and Terminal.app with fish, zsh, and bash I see a bell glyph printed on the tab if I pull a `'\a` (`\007`) while a tab/window is inactive/not getting input. It goes away when I focus it. 

Your report does remind me that I kind of am disappointed that fish itself almost under no circumstances will a BEL.Considering the general lack of richness that goes along with a terminal emulator and shell in 2015 I'm surprised there are zero situations where they'd like to flash the screen visually or ring an alert (depending on how it's configured. I prefer visual bells.) It's a very nice way to quickly indicate that an action is having no effect. 

As for titles I actually find, since fish does a good job of reporting the accurate CWD, and most terminals I use are able to figure out the active process, the most preferable thing to do with my titles bars was to turn off a custom title in fish, disabling the function. I simply pick and choose the things I'd like to see in my terminal emulator for the tab and title bar separately: current directory? full path? activity indicatos? active program, open document...  and it's a more easy, pleasant, and richer experience I'm having than I ever had trying to shove a carefully formatted string up there in the past.

**Comment by eugenesvk** [comment]
@floam I have a pencil icon instead of the bell when I run this command in `fish` on the out-of-focus tab (iTerm2/fish). 
<img width="164" alt="fish pencil instead of bell" src="https://cloud.githubusercontent.com/assets/12956286/13202419/3498d952-d8ab-11e5-891c-f99f6329125a.png">
Although I've noticed that the bell appears for a brief moment before the following progress circle appears (iTerm2/fish)
<img width="139" alt="fish progress circle" src="https://cloud.githubusercontent.com/assets/12956286/13202443/8d825060-d8ac-11e5-9978-18a4ee7f317e.png">
Also, a piece of trivia :) — in iTerm2/bash the bell doesn't disappear on focus, it only disappears after I start typing.
However, it does show up correctly in default Terminal app. So I guess it's an interaction of `fish` and `iTerm2` that's messing the bell up.

And I agree about sending the same text twice to tab and window titles.
I want to see last folder only in the tab and full path in the window title due to space constraints for the tab.

**Comment by floam** [comment]
Hm, I get do actually get the bell glyph here with iTerm Build 2.9.20160206 and fish HEAD easily right now and only the pencil if I print something without causing an alarm. However I did install iTerm's ["integration components" for fish](https://iterm2.com/misc/fish_startup.in) which are basically a handful of commands hooking into event handlers to send it information it wants over proprietary means to allow it to do some stuff. 

<img width="143" alt="screenshot 2016-02-21 at 5 10 04 am" src="https://cloud.githubusercontent.com/assets/291142/13202730/414df48a-d85a-11e5-9e50-0b6cc6724108.png">

**Comment by eugenesvk** [comment]
@floam I wasn't aware of the fish integration components (had only installed for bash/zsh).
Thanks for the tip! Otherwise also have the latest iTerm 2.9.20160206 and --HEAD version of fish.
Btw, the `sleep 1; echo -ne '\a'` gives me the same output in iTerm/fish (bell for a split second, then progress circle, then pencil).

**Comment by faho** [comment]
Is there anything left to do here or does iterm's integration solve the issue?

**Comment by eugenesvk** [comment]
@faho Yes, the main issue remains — allowing two different settings for window title and tab title
iTerm's integration didn't solve the bell issue, I still get bell->progress circle->pencil in fish (but not in bash) in iTerm.

**Comment by eugenesvk** [comment]
@floam nice, I have that status triangle on the left side of the path back with shell integration! It's much better vs coloring prompt on error via shell function since it actually shows red color on the same line as the command itself (i.e. previous line), not at the current prompt!
Thanks again for the tip
<img width="7" alt="screen shot 2016-02-21 at 6 30 02" src="https://cloud.githubusercontent.com/assets/12956286/13203432/272a4404-d8c9-11e5-8a36-f2fa4c928dc7.png">

**Comment by floam** [comment]
> Your report does remind me that I kind of am disappointed that fish itself almost under no circumstances will ring a BEL... 

So I made PR #2755 for that today for completion errors.

**Comment by eugenesvk** [comment]
Thanks, though I'd prefer the screen flashing to be optional. In iTerm there is already a very cool option that shows a flashing full-screen bell ("Flash visual bell" in Preferences->Profiles->Terminal)

**Comment by floam** [comment]
Hm. Well, if you turn that back off, your screen won't flash. I didn't change what fish does visually when there is a completion error: it still does the subtle background color change on just the typed command.

**Comment by krader1961** [comment]
> I didn't change what fish does visually when there is a completion error

That is literally true but not the whole story. Sending a bell character is a time-honored technique for signaling an error. All legacy hardware like a real VT220 will issue a bell sound. Modern emulators, like xterm and iTerm2, generally allow you to configure them so the bell character issues a bell sound, visually flash the screen, both, or neither. So, yes, people don't have to hear or see the bell character. The question is whether most fish users agree that a completion error should emit the bell character. I vote yes.

**Comment by floam** [comment]
Was just worried there may have been confusion, was trying to ensure he definitely didn't think I was adding back the many years-ago fish behavior of having it literally flash the terminal window crudely (`tput flash`). I was a little confused as was explaining that he likes the "very cool" iTerm option... which is exactly what the bell would cause.

**Comment by eugenesvk** [comment]
@floam well, I like the option, but I would hate if this option were forced on me without the ability to disable it. Used to like the sound before it became annoying and I replaced it with the visual bell.
@krader1961 I vote for a configurable variable in fish config files.

**Comment by krader1961** [comment]
@eugenesvk: I don't think this (sending a bell character on completion error) justifies being configurable. It should only be adopted if it appears the majority of fish users have no objection.  There will always be a few people who dislike a given default behavior. That does not justify making every behavior configurable. My question for you is how do you deal with all the other programs that send a bell character to the terminal to signal an important event?

**Comment by eugenesvk** [comment]
@krader1961 sorry, I didn't mean sending bell char, I meant the screen flash. But there is no flash from fish, so it's all good.
I'm fine for fish to include bell without any config.

Re. other programs I have a visual bell option turned on with a sound.
When working with bash before fish I turned the sound off because bash would send useless, but annoying multiple bells when I was holding backspace to delete a command and reached beginning of the line.

**Comment by StanBright** [comment]
I am giving a try to "fish 2.2.0" as well. However, the one last bit that bugs me is the issue with having the same Window & Tab title.

My workflow is to name every tab logically - "db console" , "tests", "server", etc. However, as long as I type any command in "fish" and the tab title (iTerm 2.9.20160313) is renamed to "fish CURRENT_DIR".

I've been looking for a solution for a few hours already, but nothing seems to help. Any idea how I could disable the tabs renaming altogether?

**Comment by StanBright** [comment]
OK, it always happens this way - you find your answer yourself immediately after asking a question :). The solution to my problem is to turn off "Terminal may set Tab/Window title" option in iTerm (Preferences > Profiles > Terminal).

**Comment by floam** [comment]
Better solution will be for fish to provide a means to set the tab title specifically. If that is not taken advantage the main terminal title function should use the escape code to overwrite both otherwise only use the one that sets the window title.

**Comment by floam** [comment]
(as it is, overwriting fish_title inhibits both tab and window title setting - and printing the CWD needlessly with fish_title when on VTE/OS X it is waiting for a directory change isn't Very Cool.)

**Comment by floam** [comment]
I'm still not exactly sure when it is this guy was seeing alerts from zsh in his window titles supposedly caused by BEL characters, but whatever it was that's out of scope for this issue - this issue is a feature request for fish to provide some means of independently setting the tab title separately from the window title.

**Comment by floam** [comment]
I'm trying to think of a way to allow users to tailor the behavior without creating new functions, and keeping it such that those who don't care about the distinction don't need to for the current behavior to stay the way it is.

I'm proposing a bit of magic, which is fish internally determining if the `fish_title` function was defined with named arguments, specifically one called `type`. This will tell `fish` the title function has opted into getting called twice each time the title is updated instead of just once. It will get called once with a first argument of `window`, and a second time with a first argument of `tab`, or whatever the most appropriate cross-terminal way of describing this is. It would be used like this:

```fish
function fish_title -a type
    switch type
        case tab
            echo stuff for the tab
        case window
            echo stuff for the window
    end
end
```

**Comment by faho** [comment]
>My current idea is a bit of magic, which is fish internally determining if the fish_title function was defined with named arguments, specifically one called type.

Why the named argument? Just pass it an argument, call it twice, ~~find out who's naughty or nice~~ and use the output? `fish_title` is usually quite simple and fast, so it shouldn't take too long.

Alternatively:

>I'm trying to think of a way to allow users to tailor the behavior without creating new functions

Why not allow specifying `fish_tab_title` and use that for the tab title if it exists, `fish_title` otherwise?

**Comment by floam** [comment]
> Why the named argument? Just pass it an argument, call it twice, find out who's naughty or nice and use the output? fish_title is usually quite simple and fast, so it shouldn't take too long.

The idea was to avoid calling executing it twice.

**Comment by floam** [comment]
Oh, and because we need a way to flag to fish whether it's output is to used for `\e]1`, `\e]2`, or `\e]0`. If we just ran it twice, it looks to us like stuff to be sent after `\e]1` and again with `\e]2`, not `\e]0` which is the current behavior.

**Comment by floam** [comment]
Here's what the parameters are, by the way:

2: Set window title
1: Sets tab title
0: Sets both window title and tab title

**Comment by floam** [comment]
The concern would be maybe there exist terminals where `\e]0` is not exactly displayed the same as sending the same string twice via `\e]1` and `\e]2`.

**Comment by floam** [comment]
Anyhow, just doing a function may be a better idea, yeah.

**Comment by floam** [comment]
Today I learned:
> The first argument to fish_title will contain the most recently executed foreground command as a string, starting with fish 2.2.

`fish_title` already gets an argument! Had no idea. And probably our default doesn't even need to run `status` as it does.

**Comment by hisnawi** [comment]
I am confused, would appreciate some help.
I am trying to set the iTerm2 window title from the command prompt using:
`echo -ne "\033]0;"Title"\007"`
But the window title directly goes back to PWD.
what I am missing?

**Comment by faho** [comment]
@hisnawi Fish already sets the title. Modify the `fish_title` function (with *just* the string you want).

**Comment by hisnawi** [comment]
You are correct. Thanks

**Comment by AdamMomen** [comment]
@hisnawi Could you share your `fish_title` function.
